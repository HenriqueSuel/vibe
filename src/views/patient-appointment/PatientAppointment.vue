<template>
  <section class="conteiner patient">
    <h1 class="patient__title">Ol√° {{ getPatient.nome }}</h1>

    <div class="card">
      <div class="flex">
        <transition name="fade" mode="out-in">
          <ListDoctor
            v-on:select-doctor="chosenDoctor = $event"
            v-if="!chosenDoctor"
          />
          <AppointementDate
            v-else-if="!done"
            v-on:previous="chosenDoctor = null"
            v-on:select-date="completeAppointment"
          />
          <section class="patient__appointment__sucess" v-else>
            <FontAwesomeIcon class="icon" :icon="faCheckCircle" />
            <h3>Agendamento feito com sucesso !</h3>
            <button @click="resetAppointment">Voltar</button>
          </section>
        </transition>
      </div>
    </div>
    <div class="card history">
      <h2>Historico</h2>

      <div class="flex">
        <date-picker
          v-model="dateHistory"
          format="DD/MM/YYYY"
          placeholder="Escolher uma data"
        ></date-picker>
        <button @click="seachHistory">Buscar</button>
      </div>

      <ul class="list">
        <li v-for="(appointment, i) of listHistory" :key="i">
          <div class="flex">
            <p class="history__value">
              {{ appointment.data | date }} - status:
              {{ appointment.status | status }}
            </p>
            <span
              v-if="appointment.status !== 'CANCELADO'"
              @click="patientCancelAppointment(appointment)"
              >Cancelar</span
            >
          </div>
        </li>
      </ul>
    </div>
  </section>
</template>

<script>
import DatePicker from 'vue2-datepicker';
import { mapGetters } from 'vuex';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { faCheckCircle } from '@fortawesome/free-solid-svg-icons';
import AppointementDate from './components/appointment-date/AppointmentDate.vue';
import ListDoctor from './components/list-doctor/ListDoctor.vue';

export default {
  components: {
    ListDoctor,
    FontAwesomeIcon,
    AppointementDate,
    DatePicker,
  },
  data() {
    return {
      chosenDoctor: null,
      done: false,
      faCheckCircle,
      dateHistory: new Date(),
      listHistory: [],
    };
  },
  computed: {
    ...mapGetters('PatientStore', ['getPatient']),
    ...mapGetters('DoctorStore', ['findDoctorByName']),
    ...mapGetters('AppointmentStore', ['getAllPatientAppointment']),
  },
  methods: {
    selectDoctor(doctor) {
      this.chosenDoctor = doctor;
      this.listDoctor = [];
      this.nameDoctor = '';
    },
    async completeAppointment(date) {
      this.done = await this.$store.dispatch('AppointmentStore/makeAppointment', {
        date,
        doctor: this.chosenDoctor,
      });
    },
    resetAppointment() {
      this.done = false;
      this.chosenDoctor = null;
    },
    seachHistory() {
      this.listHistory = this.getAllPatientAppointment(
        this.getPatient.id,
        this.dateHistory,
      );
    },
    async patientCancelAppointment(appointment) {
      await this.$store.dispatch('AppointmentStore/cancelAppointment', appointment);
      this.listHistory = [];
    },
  },
};
</script>

<style scoped lang="scss">
.conteiner {
  padding: 32px;
}

.patient {
  &__title {
    margin-bottom: 16px;
  }
  &__card-title {
    font-size: 16px;
    font-weight: 100;
  }
  &__appointment {
    display: flex;
    justify-content: center;
    width: 100%;
    flex-direction: column;
    align-items: center;

    &__sucess {
      width: 100%;
      display: flex;
      flex-direction: column;
      align-items: center;
      h3 {
        margin-top: 16px;
        font-weight: 100;
      }
      .icon {
        font-size: 300px;
        color: var(--sucess);
      }
    }
  }
}

.history {
  .list {
    li {
      cursor: default;
    }
  }
  &__value {
    flex-grow: 1;
  }
  span {
    cursor: pointer;
  }
}
button {
  border: 0px;
  margin: 16px;
  color: white;
  background: #e72d51;
  border-radius: 10px;
  font-size: 16px;
  line-height: 21px;
  padding: 16px;
  display: flex;
}
</style>
