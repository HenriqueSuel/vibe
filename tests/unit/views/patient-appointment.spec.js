import { shallowMount, createLocalVue } from '@vue/test-utils';
import { VueMaskDirective } from 'v-mask';
import Vuex from 'vuex';
import flushPromises from 'flush-promises';

import PatientAppointment from '@/views/patient-appointment/PatientAppointment.vue';
import Vuelidate from 'vuelidate';

const localVue = createLocalVue(Vuelidate);
localVue.use(Vuex);
localVue.use(Vuelidate);

describe('Login.vue', () => {
  const doctor = { id: '99272cf5-bd9c-4413-b6bf-124ebb45ec80', nome: 'Henrique Suel', crm: '1' };
  const appointment = {
    id: '239fdc6c-d8ef-49ef-859e-9fe66b2d3833', data: '2020-09-26T15:00:00.000Z', idUser: '99272cf5-bd9c-4413-b6bf-124ebb45ec80', nameUser: 'Henrique', status: 'AGENDADO', cid: '', descricao: '',
  };
  let actions;
  let store;
  let getters;

  beforeEach(() => {
    actions = {
      makeAppointment: jest.fn().mockResolvedValue(true),
      cancelAppointment: jest.fn().mockResolvedValue(true),
    };

    getters = {
      getAllPatientAppointment: () => [{ id: '1' }],
      getPatient: jest.fn().mockResolvedValue({ date: new Date(), nameUser: 'henrique' }),
      findDoctorByName: jest.fn().mockResolvedValue({ date: new Date(), nameUser: 'henrique' }),
    };

    store = new Vuex.Store({
      state: {},
      modules: {
        PatientStore: {
          namespaced: true,
          actions,
          getters,
        },
        DoctorStore: {
          namespaced: true,
          actions,
        },
        AppointmentStore: {
          namespaced: true,
          actions,
          getters,
        },
      },
    });
  });

  it('should set a value for the doctor', () => {
    const wrapper = shallowMount(PatientAppointment, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    wrapper.vm.selectDoctor(doctor);
    expect(wrapper.vm.$data.chosenDoctor.id).toEqual(doctor.id);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should set a value for the doctor', async () => {
    const wrapper = shallowMount(PatientAppointment, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    wrapper.vm.completeAppointment(new Date());
    await flushPromises();
    expect(actions.makeAppointment).toHaveBeenCalled();
  });
  it('Should be reset as variables', () => {
    const wrapper = shallowMount(PatientAppointment, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    wrapper.setData({ done: true, chosenDoctor: doctor });
    wrapper.vm.resetAppointment();
    expect(wrapper.vm.$data.done).toBeFalsy();
    expect(wrapper.vm.$data.chosenDoctor).toEqual(null);
  });

  it('Should cancel the appointment', async () => {
    const wrapper = shallowMount(PatientAppointment, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    wrapper.setData({ done: true, chosenDoctor: doctor });
    wrapper.vm.patientCancelAppointment(appointment);
    await flushPromises();
    expect(actions.cancelAppointment).toHaveBeenCalled();
  });
});
