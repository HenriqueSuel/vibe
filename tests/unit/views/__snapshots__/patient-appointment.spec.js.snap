// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue should set a value for the doctor 1`] = `
<section class="conteiner patient">
  <h1 class="patient__title">Ol√° </h1>
  <div class="card">
    <div class="flex">
      <transition-stub name="fade" mode="out-in">
        <listdoctor-stub></listdoctor-stub>
      </transition-stub>
    </div>
  </div>
  <div class="card history">
    <h2>Historico</h2>
    <div class="flex">
      <date-picker-stub value="Sun Sep 27 2020 23:43:58 GMT-0300 (GMT-03:00)" defaultvalue="Sun Sep 27 2020 00:00:00 GMT-0300 (GMT-03:00)" disableddate="function _default() {
          return false;
        }" type="date" getclasses="function _default() {
          return [];
        }" titleformat="YYYY-MM-DD" format="DD/MM/YYYY" timetitleformat="YYYY-MM-DD" disabledtime="function _default() {
          return false;
        }" hourstep="1" minutestep="1" secondstep="1" scrollduration="100" valuetype="date" rangeseparator=" ~ " placeholder="Escolher uma data" editable="true" clearable="true" prefixclass="mx" inputclass="mx-input" inputattr="[object Object]" appendtobody="true" popupstyle="[object Object]" confirmtext="OK" shortcuts=""></date-picker-stub> <button>Buscar</button>
    </div>
    <ul class="list"></ul>
  </div>
</section>
`;
